File: examples/csharp/Player/Agg/Handlers/ReleaseHandler.cs
using Google.Protobuf.WellKnownTypes;
using Angzarr.Client;
using Angzarr.Examples;

namespace Player.Agg.Handlers;

/// <summary>
/// Handler for ReleaseFunds command.
/// </summary>
public static class ReleaseHandler
{
    public static FundsReleased Handle(ReleaseFunds cmd, PlayerState state)
    {
        // Guard
        if (!state.Exists)
            throw CommandRejectedError.PreconditionFailed("Player does not exist");

        // Validate
        var tableKey = Convert.ToHexString(cmd.TableRoot.ToByteArray()).ToLowerInvariant();
        if (!state.TableReservations.TryGetValue(tableKey, out var reservedForTable) || reservedForTable == 0)
            throw CommandRejectedError.PreconditionFailed("No funds reserved for this table");

        // Compute
        var newReserved = state.ReservedFunds - reservedForTable;
        var newAvailable = state.Bankroll - newReserved;
        return new FundsReleased
        {
            Amount = new Currency { Amount = reservedForTable, CurrencyCode = "CHIPS" },
            TableRoot = cmd.TableRoot,
            NewAvailableBalance = new Currency { Amount = newAvailable, CurrencyCode = "CHIPS" },
            NewReservedBalance = new Currency { Amount = newReserved, CurrencyCode = "CHIPS" },
            ReleasedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
    }
}

File: examples/cpp/player/agg/handlers/release_handler.cpp
#include "release_handler.hpp"
#include "angzarr/errors.hpp"
#include <chrono>
#include <iomanip>
#include <sstream>

namespace player {
namespace handlers {

namespace {
std::string bytes_to_hex(const std::string& bytes) {
    std::ostringstream ss;
    for (unsigned char c : bytes) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(c);
    }
    return ss.str();
}
} // anonymous namespace

examples::FundsReleased handle_release(const examples::ReleaseFunds& cmd, const PlayerState& state) {
    // Guard
    if (!state.exists()) {
        throw angzarr::CommandRejectedError::precondition_failed("Player does not exist");
    }

    std::string table_key = bytes_to_hex(cmd.table_root());
    auto it = state.table_reservations.find(table_key);
    if (it == state.table_reservations.end() || it->second == 0) {
        throw angzarr::CommandRejectedError::precondition_failed("No funds reserved for this table");
    }

    // Compute
    int64_t reserved_for_table = it->second;
    int64_t new_reserved = state.reserved_funds - reserved_for_table;
    int64_t new_available = state.bankroll - new_reserved;

    examples::FundsReleased event;
    event.mutable_amount()->set_amount(reserved_for_table);
    event.mutable_amount()->set_currency_code("CHIPS");
    event.set_table_root(cmd.table_root());
    event.mutable_new_available_balance()->set_amount(new_available);
    event.mutable_new_available_balance()->set_currency_code("CHIPS");
    event.mutable_new_reserved_balance()->set_amount(new_reserved);
    event.mutable_new_reserved_balance()->set_currency_code("CHIPS");

    auto now = std::chrono::system_clock::now();
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    event.mutable_released_at()->set_seconds(seconds);

    return event;
}

} // namespace handlers
} // namespace player
